File: src/lexer.c
#pragma once
#include "token.c"
#include "vector.c"
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define IS_NUMBER(n) (n > 47 && n < 58)
#define IS_SPECIAL_CHARACTER(n) (strchr("{}()[]:.,;=*/%+-<>", n) != NULL)
#define IS_IDENTIFIER_BEGIN(n) ((n > 96 && n < 123) || (n > 64 && n < 91) || n == '_' || n == '#')
#define IS_IDENTIFIER(n) (IS_IDENTIFIER_BEGIN(n) || IS_NUMBER(n))
#define DO_IGNORE(n) (strchr(" \n\t\b\r", n) != NULL)
#define BUF_SIZE 32

static const char *const KEYWORDS[] = {
  "var", "const",
  "if", "else", "for", "while", "switch", "case", "where",
  "fn", "struct", "error"
};
static const char *const OPERATORS[] = {
  "+", "-", "*", "/", "%", "&", "|", "^", "!", "=", "?", "<", ">", ":",
  "++", "--", "+=", "-=", "*=", "/=", "%=", "&=", "|=", "^=", "!=", "==", "<<", ">>", "|>", "->",
  "<<=", ">>=", "|>="
};
static const char *const TYPES[] = {
  "Signed", "Unsigned", "Float", "Array", "Function",
  "Type", "Error"
};

static const size_t KEYWORDS_SIZE = sizeof(KEYWORDS) / sizeof(KEYWORDS[0]);
static const size_t OPERATORS_SIZE = sizeof(OPERATORS) / sizeof(OPERATORS[0]);
static const size_t TYPES_SIZE = sizeof(TYPES) / sizeof(TYPES[0]);

typedef struct {
  FILE *fptr;
  char *error;
  size_t current_line;
  size_t index;
  char current_char;
  bool eof;
  char buf[BUF_SIZE];
} Lexer;

void advance(Lexer *const restrict stream, size_t ammount) {
  if (ammount > 1) {
    for (size_t i = 0; i < ammount; ++i)
      advance(stream, 1);
  } else {
    ++stream->index;
    if (stream->index == BUF_SIZE - 1 || stream->buf[stream->index - 1] == '\n') {
      stream->eof = !fgets(stream->buf, BUF_SIZE, stream->fptr);
      stream->index = 0;
    }
  }

  stream->current_char = stream->buf[stream->index];
  stream->current_line += stream->current_char == '\n';
}

// PERF: O(n), can be improved.
bool isIn(Token *const restrict token, const char *const array[], size_t length) {
  for (size_t i = 0; i < length; ++i)
    if (strcmp(array[i], token->value.data) == 0)
      return true;
  return false;
}

void getIdentifier(Lexer *const restrict stream, Token *const restrict token) {
  while (IS_IDENTIFIER(stream->current_char)) {
    vectorAdd(&token->value, &stream->current_char);
    advance(stream, 1);
  }

  token->type = IDENTIFIER;
}

void getOperator(Lexer *const restrict stream, Token *const restrict token) {
  for (;;) {
    vectorAdd(&token->value, &stream->current_char);
    if (isIn(token, OPERATORS, OPERATORS_SIZE))
      advance(stream, 1);
    else
      break;
  }

  token->type = OPERATOR;
  vectorPop(&token->value);
}

void getNumber(Lexer *const restrict stream, Token *const restrict token) {
  char dotcount = 0;
  while (IS_NUMBER(stream->current_char) || stream->current_char == '.') {
    if (stream->current_char == '.')
      if (++dotcount > 1)
        break;

    vectorAdd(&token->value, &stream->current_char);
    advance(stream, 1);
  }

  token->type = dotcount ? FLOAT_LITERAL : INT_LITERAL;
}

void getString(Lexer *const restrict stream, Token *const restrict token) {
  advance(stream, 1);
  while (stream->current_char != '"' && !stream->eof) {
    vectorAdd(&token->value, &stream->current_char);
    advance(stream, 1);
  }
  advance(stream, 1);

  token->type = STRING_LITERAL;
}

void getCharacter(Lexer *const restrict stream, Token *const restrict token) {
  advance(stream, 1);
  vectorAdd(&token->value, &stream->current_char);
  advance(stream, 1);

  if (stream->current_char != '\'') {
    char *buffer = calloc(64, sizeof(char));
    snprintf(buffer, 64 * sizeof(char), "Expected closing quote ('). [line %zu]", stream->current_line);
    stream->error = buffer;
    stream->eof = true;
  }
  advance(stream, 1);

  token->type = CHAR_LITERAL;
}

void lexerInit(Lexer *const restrict stream, char *const restrict file_name) {
  stream->fptr = fopen(file_name, "r");
  stream->error = NULL;
  stream->current_line = 1;
  stream->index = 0;

  if (!fgets(stream->buf, BUF_SIZE, stream->fptr))
    stream->eof = true;
  stream->current_char = stream->buf[0];
  stream->eof = stream->eof ? true : stream->current_char == '\0';
}

void lexerDeinit(Lexer *const restrict stream) {
  free(stream->error);
  fclose(stream->fptr);
  stream->error = NULL;
  stream->fptr = NULL;
}

size_t lexerTokenize(Lexer *const restrict stream, Token *const restrict buf, size_t buffer_size) {
  if (stream->eof)
    return 0;

  Token token;
  size_t token_index = 0;
  while (token_index < buffer_size && !stream->eof) {
    printf("loop %zu '%c'\n", token_index, stream->current_char);
    if (DO_IGNORE(stream->current_char)) {
      advance(stream, 1);
      if (stream->eof)
        break;
      continue;
    }

    tokenInit(&token);
    token.line = stream->current_line;
    token.begin = stream->index;
    if (IS_IDENTIFIER_BEGIN(stream->current_char)) {
      getIdentifier(stream, &token);
      if (isIn(&token, KEYWORDS, KEYWORDS_SIZE))
        token.type = KEYWORD;
      else if (isIn(&token, TYPES, TYPES_SIZE))
        token.type = TYPE;
    } else if (IS_SPECIAL_CHARACTER(stream->current_char)) {
      getOperator(stream, &token);
    } else if (IS_NUMBER(stream->current_char)) {
      getNumber(stream, &token);
    } else if (stream->current_char == '"') {
      getString(stream, &token);
    } else if (stream->current_char == '\'') {
      getCharacter(stream, &token);
    } else {
      tokenDeinit(&token);
      char *buffer = calloc(64, sizeof(char));
      snprintf(buffer, sizeof(char) * 64, "Couldn't recognize token. [line %zu]", stream->current_line);
      stream->error = buffer;
      printf("return %s\n", stream->error);
      return token_index;
    }
    token.end = stream->index - 1;
    vectorTrim(&token.value);
    buf[token_index] = token;
    ++token_index;
  }

  printf("return %zu\n", token_index);
  return token_index;
}

File: src/lexer.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

const TokenKind = enum(u4) {
    KEYWORD,
    IDENTIFIER,
    INTEGER,
    FLOAT,
    STRING,
};

const TokenValue = union(enum) {
    int: u64,
    float: f64,
    string: std.ArrayList(u8),
};

const Token = struct {
    kind: TokenKind,
    value: TokenValue,
    line: u32,
    begin: u32,
    end: u32,
};

const LexerError = error{
    UnexpectedCharacter,
    NumberTooBig,
};

pub fn Lexer() type {
    return struct {
        const Self = @This();
        index: u32 = 0,
        line: u32 = 1,
        char_index: u32 = 1,
        char: u8,
        eof: bool = false,
        code: *[]const u8,
        allocator: Allocator,
        keywords: std.StringHashMap(bool),

        pub fn init(allocator: Allocator, code: *[]const u8) !Self {
            var self = Self{
                .char = code.*[0],
                .code = code,
                .allocator = allocator,
                .keywords = std.StringHashMap(bool).init(allocator),
            };

            try self.keywords.put("if", true);
            try self.keywords.put("else", true);
            return self;
        }

        fn advance(self: *Self) void {
            self.index += 1;
            self.char_index += 1;
            self.eof = self.index >= self.code.*.len;
            self.char = if (self.eof) 0 else self.code.*[self.index];
            if (self.char == '\n') {
                self.line += 1;
                self.char_index = 1;
            }
        }

        fn isKeyword(self: *Self, identifier: []u8) bool {
            return self.keywords.get(@as([]const u8, identifier)) orelse false;
        }

        fn isIdentifier(self: *Self) bool {
            return switch (self.char) {
                'A'...'Z', 'a'...'z', '#', '_', '$' => true,
                else => false,
            };
        }

        fn getIdentifier(self: *Self) Allocator.Error!Token {
            const begin = self.char_index;
            var str = std.ArrayList(u8).init(self.allocator);
            errdefer str.deinit();

            while (self.isIdentifier()) {
                try str.append(self.char);
                self.advance();
            }

            str.shrinkAndFree(str.items.len);
            return Token{ .kind = if (self.isKeyword(str.items)) .KEYWORD else .IDENTIFIER, .value = .{ .string = str }, .line = self.line, .begin = begin, .end = self.index };
        }

        fn isNumber(self: *Self) bool {
            return switch (self.char) {
                '0'...'9', '.' => true,
                else => false,
            };
        }

        fn getNumber(self: *Self) LexerError!Token {
            const begin = self.char_index;
            var num: u64 = 0;
            var den: u6 = 1;
            var dotcount: u2 = 0;

            while (self.isNumber()) {
                if (self.char == '.') {
                    dotcount += 1;
                    if (dotcount > 1)
                        break;
                    self.advance();
                    continue;
                }

                num *= 10;
                try {
                    num += self.char - '0';
                } catch return LexerError.NumberTooBig;
                if (dotcount > 0)
                    den *= 10;
                self.advance();
            }

            if (dotcount > 0) {
                return Token{ .kind = .FLOAT, .value = .{ .float = @as(f64, @floatFromInt(num)) / @as(f64, @floatFromInt(den)) }, .line = self.line, .begin = begin, .end = self.index };
            }
            return Token{ .kind = .INTEGER, .value = .{ .int = num }, .line = self.line, .begin = begin, .end = self.index };
        }

        fn getString(self: *Self, char: u8) (Allocator.Error || LexerError)!Token {
            const line = self.line;
            const begin = self.char_index;
            var str = std.ArrayList(u8).init(self.allocator);
            errdefer str.deinit();

            self.advance();
            while (self.char != char) {
                if (self.eof)
                    return LexerError.UnexpectedCharacter;
                try str.append(self.char);
                self.advance();
            }
            self.advance();

            str.shrinkAndFree(str.items.len);
            return Token{ .kind = .STRING, .value = .{ .string = str }, .line = line, .begin = begin, .end = self.index };
        }

        fn doIgnore(self: *Self) bool {
            return switch (self.char) {
                ' ', '\n', '\t', '\r' => true,
                else => false,
            };
        }

        pub fn nextToken(self: *Self) (Allocator.Error || LexerError)!?Token {
            while (self.doIgnore()) {
                self.advance();
            }
            if (self.eof) {
                return null;
            }

            var token = (try self.allocator.create(Token)).*;
            errdefer self.allocator.destroy(&token);

            if (self.isNumber()) {
                token = try self.getNumber();
            } else if (self.isIdentifier()) {
                token = try self.getIdentifier();
            } else if (self.char == '"' or self.char == '\'') {
                token = try self.getString(self.char);
            } else {
                return LexerError.UnexpectedCharacter;
            }

            return token;
        }
    };
}
